### ISSUE TITLE
Can't mount volume from single replica.

### ISSUE THREAD

**Issue by Dudoleitor** [longhorn/longhorn#3817]
Hi,
I'm having problems trying to recover data from a broken volume. Due to issue #3576 I wasn't able to properly configure backups and I'm backing up the /var/lib/longhorn folder as a temporary fix. This allowed me to recover previous snapshots of the folder /var/lib/longhorn/broken-volume which contains volume.meta, volume-head etc; to work in a clean enviroment I always copy the folder to a new dir and work there.

## Steps followed

Following this guide [data-recovery/export-from-replica](https://longhorn.io/docs/1.2.0/advanced-resources/data-recovery/export-from-replica/) I tried to mount the volume but there seems to be something wrong.
This are the steps I follow:
- Copy the recovered folder /var/lib/longhorn/broken-volume to /root/volume
- Open volume.meta to check the volume size, eg 2147483648
- Run `docker run -it -v /dev:/host/dev -v /proc:/host/proc -v root/volume:/volume --privileged longhornio/longhorn-engine:v1.2.3 launch-simple-longhorn volume-recover 2147483648` (in a separate shell)
- Mount the folder using `mount /dev/longhorn/volume-recover /mnt` or `mount /dev/sdX /mnt`
- Opening the volume with `ls -la /mnt` which returns `ls: reading directory '/mnt': Bad message`

While the container is running, using `lsblk -f` I see the line corresponding to the volume, the file system is correctly recognised as ext4 but the columns `FSAVAIL` `FSUSE%` are empty.

I can't figure out why it's not working... I even tried to use a healthy volume copied directly from the longhorn instance but I'm still not able to mount a replica.

## Logs

If someone has the patience to read it, here I posted the full log from the docker container: [gist.github.com](https://gist.github.com/Dudoleitor/8fdddd2d16f6f5a0b78513794c4f1130). The only line that seems suspect to me is `Failed to initialize RDMA; load kernel modules?` (34) but the execution procedes after that.

## Environment

 - Longhorn engine image (used by the cluster): longhornio/longhorn-engine:v1.2.3
 - Longhorn engine image (used in the tests): longhornio/longhorn-engine:v1.2.3
 - Kubernetes k3s version: v1.22.7+k3s1
 - Rancher version: v2.6.3 
 - Node config
   - OS type and version: Debian buster 5.10.0-13-amd64
   - CPU per node: 6
   - Memory per node: 4GB
   - Network bandwidth and latency between the nodes: minimum bandwith 1Gbps, latency < 1ms
 - Underlying Infrastructure: Baremetal, nodes are VM running on proxmox

**Comment by Dudoleitor** [comment]
Hi,
first of all thanks for your interest.
To be sure I wasn't working with a failed filesystem, I tried to open a volume copied directly from a working deployment and the issue still came up. To test the behaviour in this situation, I was sure the pod was running ok and I even tried to redeploy the workload. The pod succesfully start, without errors, and the volume was re-attached correctly.

Running mount does succeed, I get no errors indeed with `lsblk` the partition is currectly listed as mounted to /mnt. As I said, after that there is no way to access the files (`ls -la /mnt` returns `ls: reading directory '/mnt': Bad message`).

Speaking of the pod I was trying to recover data for, its volume was completely wiped and lost all the data inside (don't know why), so I needed to recover data from a backup. As a matter of facts I managed to recover the data I lost in another way, so this issue is not urgent for me.

Anyway I'd like to understand what's going on and to know how to recover data from the replica of a volume. I think this could be an issue related to launch-simple-longhorn, so I'm available to help understand if something doesn't work or it's just me who is doing something wrong.

**Comment by shuo-wu** [comment]
I just tried something similar and it works for me:
1. Create and attach a volume, then make an EXT4 filesystem for the volume.
2. Detach the volume so that there is nothing using the underlying replica data directory.
3. Copy the content of the replica directory to another random directory.
4. Run the command with a random volume name. e.g., `docker run -it -v /dev:/host/dev -v /proc:/host/proc -v /mnt/replica-random-dir:/volume --privileged longhornio/longhorn-engine:v1.2.3 launch-simple-longhorn volume-random-name 2147483648`.
5. Mount the volume to a random mount point in another shell session.

You said that `I tried to open a volume copied directly from a working deployment and the issue still came up.`. I am not sure if the cache in the working workload would affect the integrity of the filesystem content. To test it, I would suggest using the directory of a detached volume and making sure the data is flushed before the detachment.

BTW, have you tried to run `fsck` for this block device?

**Comment by Dudoleitor** [comment]
Thanks for your interest!
I tried to open a volume copied directly from a running deployment and now it works... I used the same deployment as the previous tests.
After that, I tried to recover the same volume from the snapshot taken by proxmox during the night and the error is still there.

I tried with another deployment, which has read write many volumes, and there were no errors mounting the volume copied directly from a running replica. Extracting the volume from the backup brought problems in this case too, when trying to mount I get `mount: /mnt: wrong fs type, bad option, bad superblock on /dev/longhorn/volume-recover, missing codepage or helper program, or other error`.

When using `fsck` I remember that it "fixed" the filesytem wiping out everything and leaving an empty partition (with just the lost+found folder).

So I'm lead to say that something brokes when proxmox captures snapshots of running replicas. Anyway, this is not the proper way to backup longhorn volumes so we can assume this isn't an issue related to launch-simple-longhorn.
